src/category_theory/monoidal/braided.lean

```diff
@@ -283,4 +299,364 @@ def discrete.braided_functor (F : M →* N) : braided_functor (discrete M) (disc
 
 end comm_monoid
 
+section tensor
+
+open free_monoidal_category
+
+def tensor_μ (X Y : C × C) : (tensor C).obj X ⊗ (tensor C).obj Y ⟶ (tensor C).obj (X ⊗ Y) :=
+(α_ X.1 X.2 (Y.1 ⊗ Y.2)).hom ≫ (𝟙 X.1 ⊗ (α_ X.2 Y.1 Y.2).inv) ≫
+(𝟙 X.1 ⊗ ((β_ X.2 Y.1).hom ⊗ 𝟙 Y.2)) ≫
+(𝟙 X.1 ⊗ (α_ Y.1 X.2 Y.2).hom) ≫ (α_ X.1 Y.1 (X.2 ⊗ Y.2)).inv
+
+lemma tensor_μ_def₁ (X₁ X₂ Y₁ Y₂ : C) :
+    tensor_μ C (X₁, X₂) (Y₁, Y₂) ≫ (α_ X₁ Y₁ (X₂ ⊗ Y₂)).hom ≫ (𝟙 X₁ ⊗ (α_ Y₁ X₂ Y₂).inv)
+  = (α_ X₁ X₂ (Y₁ ⊗ Y₂)).hom ≫ (𝟙 X₁ ⊗ (α_ X₂ Y₁ Y₂).inv) ≫ (𝟙 X₁ ⊗ ((β_ X₂ Y₁).hom ⊗ 𝟙 Y₂)) :=
+by { dsimp [tensor_μ], simp }
+
+lemma tensor_μ_def₂ (X₁ X₂ Y₁ Y₂ : C) :
+    (𝟙 X₁ ⊗ (α_ X₂ Y₁ Y₂).hom) ≫ (α_ X₁ X₂ (Y₁ ⊗ Y₂)).inv ≫ tensor_μ C (X₁, X₂) (Y₁, Y₂)
+  = (𝟙 X₁ ⊗ ((β_ X₂ Y₁).hom ⊗ 𝟙 Y₂)) ≫ (𝟙 X₁ ⊗ (α_ Y₁ X₂ Y₂).hom) ≫ (α_ X₁ Y₁ (X₂ ⊗ Y₂)).inv :=
+by { dsimp [tensor_μ], simp }
+
+lemma tensor_μ_natural {X₁ X₂ Y₁ Y₂ U₁ U₂ V₁ V₂ : C}
+  (f₁ : X₁ ⟶ Y₁) (f₂ : X₂ ⟶ Y₂) (g₁ : U₁ ⟶ V₁) (g₂ : U₂ ⟶ V₂) :
+  ((f₁ ⊗ f₂) ⊗ (g₁ ⊗ g₂)) ≫ tensor_μ C (Y₁, Y₂) (V₁, V₂) =
+    tensor_μ C (X₁, X₂) (U₁, U₂) ≫ ((f₁ ⊗ g₁) ⊗ (f₂ ⊗ g₂)) :=
+begin
+  dsimp [tensor_μ],
+  slice_lhs 1 2 { rw [associator_naturality] },
+  slice_lhs 2 3 { rw [←tensor_comp,
+                      comp_id f₁, ←id_comp f₁,
+                      associator_inv_naturality,
+                      tensor_comp] },
+  slice_lhs 3 4 { rw [←tensor_comp, ←tensor_comp,
+                      comp_id f₁, ←id_comp f₁,
+                      comp_id g₂, ←id_comp g₂,
+                      braiding_naturality,
+                      tensor_comp, tensor_comp] },
+  slice_lhs 4 5 { rw [←tensor_comp,
+                      comp_id f₁, ←id_comp f₁,
+                      associator_naturality,
+                      tensor_comp] },
+  slice_lhs 5 6 { rw [associator_inv_naturality] },
+  simp only [assoc],
+end
+
+lemma tensor_left_unitality (X₁ X₂ : C) :
+    (λ_ (X₁ ⊗ X₂)).hom
+  = ((λ_ (𝟙_ C)).inv ⊗ 𝟙 (X₁ ⊗ X₂)) ≫
+    tensor_μ C (𝟙_ C, 𝟙_ C) (X₁, X₂) ≫
+    ((λ_ X₁).hom ⊗ (λ_ X₂).hom) :=
+begin
+  dsimp [tensor_μ],
+  have :
+      ((λ_ (𝟙_ C)).inv ⊗ 𝟙 (X₁ ⊗ X₂)) ≫
+      (α_ (𝟙_ C) (𝟙_ C) (X₁ ⊗ X₂)).hom ≫
+      (𝟙 (𝟙_ C) ⊗ (α_ (𝟙_ C) X₁ X₂).inv)
+    = 𝟙 (𝟙_ C) ⊗ ((λ_ X₁).inv ⊗ 𝟙 X₂),
+  { let lhs :=
+      ((λ_ unit).inv ⊗ 𝟙 ((of X₁) ⊗ (of X₂))) ≫
+      (α_ unit unit ((of X₁) ⊗ (of X₂))).hom ≫
+      (𝟙 unit ⊗ (α_ unit (of X₁) (of X₂)).inv),
+    let rhs :=
+      𝟙 unit ⊗ ((λ_ (of X₁)).inv ⊗ 𝟙 (of X₂)),
+    change project_map id _ _ lhs = project_map id _ _ rhs,
+    congr },
```

Comment:
This is probably worth calling out because I'm using this trick a lot.  (I was quite excited when I discovered it, so please forgive me if it's something you already know well.)

Namely, I'm using the coherence theorem proved in category_theory.monoidal.free.coherence to conclude that two given morphisms in a monoidal category composed out of associators, unitors, identities, and their tensor products are equal if their source and target objects are equal.  The proofs have the same structure: We lift the sides of the equation in question to morphisms in the free monoidal category `F C` that are mapped back to the original morphisms by the projection functor from `F C` to `C`.  The liftings are equal by the coherence theorem (which asserts that Hom-s in the free monoidal category are subsingletons), and hence their images are also equal.  The lifting is a completely mechanical procedure and produces liftings whose projections are definitionally equal to the original morphisms.  We take basically the same formulas for the morphisms but replace every object with its lift: the unit object `𝟙_ C` with `unit`, an object `X` with `of X`, tensor product `X ⊗ Y` with `(of X) ⊗ (of Y)` etc.

I have a feeling that it should be possible to automate these kinds of proofs.  One approach I briefly considered is to use type classes to automatically build liftings similarly to how `is_iso` is used to build inverses.  Naively, we would define a type class say `is_coh` containing a lifting of a given morphism and a proof that it is a lifting, and define instances for associator and unitors, and chained instances for tensor products and compositions.  The problem I ran into was lifting of objects: it is important that it distributes over tensor product and that the unit object `𝟙_ C` is mapped to the unit object `unit` of `F C` and not to `of (𝟙_ C)`.  I'm not sure how to define this kind of mapping.  Perhaps again with type classes and overlapping instances (I know it's possible to assign priorities to instances to guide instance search), but I'm not sufficiently familiar with type classes in Lean yet to see if it's possible and if so, how.

Another option / question: is it possible to write a tactic that would perform the kind of lifting I described?  (I don't know anything yet about writing tactics in Lean.)

Reply:
@manzyuk, I took the liberty of pushing a commit that uses the typeclass approach you suggested. Could you take a look at it and see if you like it? If so it should be straightforward to write a tactic that does the final piece of boilerplate.

Reply:
I wrote the tactic, too.

Reply:
This is great!  Let me try and confirm that the tactic can replace all manual proofs I have in this and subsequent PRs.

Reply:
Since I've now cleaned up the tactic, and made it more powerful, in a separate PR, I've removed my commits from this PR. Sorry for the noise.

I'm pretty sure we will want to wait for #13125 to be merged, and then update this to use it, however.

Reply:
Makes sense.  Thanks for working on this and adding the tactic!